server/src/main/java/org/candlepin/model/Transactional.java

```diff
@@ -0,0 +1,301 @@
+     * Multiple listeners may be added, and a given listener may be added multiple times.
+     *
+     * @param listener
+     *  the transaction listener to trigger on rollback
+     *
+     * @throws IllegalArgumentException
+     *  if listener is null
+     *
+     * @return
+     *  this transactional wrapper
+     */
+    public Transactional<O> onRollback(Listener listener) {
+        if (listener == null) {
+            throw new IllegalArgumentException("listener is null");
+        }
+
+        this.rollbackListeners.add(listener);
+        return this;
+    }
+
+    /**
+     * Adds a supplemental action to perform upon completion of the transaction -- successful or
+     * otherwise. Multiple listeners may be added, and a given listener may be added multiple times.
+     *
+     * @param listener
+     *  the transaction listener to trigger on completion
+     *
+     * @throws IllegalArgumentException
+     *  if listener is null
+     *
+     * @return
+     *  this transactional wrapper
+     */
+    public Transactional<O> onComplete(Listener listener) {
+        this.onCommit(listener);
+        this.onRollback(listener);
+
+        return this;
+    }
+
+    /**
+     * Sets this wrapper to automatically commit the transaction if an uncaught exception occurs.
+     * By default, the wrapper will rollback the transaction on exception, but this method can be
+     * used to specify the behavior.
+     *
+     * @return
+     *  this transactional wrapper
+     */
+    public Transactional<O> commitOnException() {
+        this.commitOnException = true;
+        return this;
+    }
+
+    /**
+     * Commits the transaction and notifies listeners
+     */
+    private void commitTransaction(EntityTransaction transaction) {
+        transaction.commit();
+
+        for (Listener listener : this.commitListeners) {
+            listener.handleTransaction();
+        }
+    }
+
+    /**
+     * Rolls back the transaction and notifies listeners
+     */
+    private void rollbackTransaction(EntityTransaction transaction) {
+        transaction.rollback();
+
+        for (Listener listener : this.rollbackListeners) {
+            listener.handleTransaction();
+        }
+    }
+
+    /**
+     * Executes this transactional, passing the specified arguments through to the action.
+     *
+     * @param args
+     *  the args to pass through to the action
+     *
+     * @return
+     *  the output of the action
+     */
+    public O execute(Object... args) {
+        if (this.action == null) {
+            throw new IllegalStateException("no action provided");
+        }
+
+        O output = null;
+        EntityTransaction transaction = null;
+
+        try {
+            transaction = this.entityManager.getTransaction();
+
+            if (!transaction.isActive()) {
+                transaction.begin();
+            }
+            else {
+                log.warn("Transaction already started");
```

Comment:
Is warning here enough? Someone using this feature could assume that the action is happening in its own transaction, and not within a transaction that is being continued from before. I am not sure what the proper way to handle this is. We could try to commit the existing transaction here, before opening a new one, but that feels wrong. The cleanest would probably be to throw an exception. Another choice would be to make this a choice on the API (have a `reuseActiveTransaction` method, while the default behaviour would be to throw an exception instead). 

Whatever we choose, we should document this in Javadoc.

Reply:
Yeah, I was leaning toward the latter, but opted to skip it for now due to wanting to get this done quickly. I'll add something here to control that a bit more.